### DISCUSSION TITLE
How to chain ops jobs and asset jobs ?

### DISCUSSION THREAD

**Discussion by eromoe** [dagster-io/dagster#15697] [Q&A]
I have a job download data first 
```python 
@job
def download_stock():
    cmds = clear_cmds(cmds_str)
    names = [extract_name(i) for i in cmds]
    for n,c in zip(names, cmds):
        # avoid name duplicate, add download prefix
        create_shell_command_op(c, name=f'download_{n}')()

```
The download logic is written in another lib.

After this the download finished, I need start build an asset for example `stock_adj_df` as below:

```python
@asset
def stock_df()->pd.DataFrame:
    return TushareRouter['stock_a'].read()

@asset
def adj_factor()->pd.DataFrame:
    return TushareRouter['adj_factor'].read()

@asset
def stock_adj_df(stock_df:pd.DataFrame, adj_factor:pd.DataFrame)->pd.DataFrame:
    stock_adj_df = pd.merge(stock_df, adj_factor, on=['ts_code', 'trade_date'], how='left')
    return stock_adj_df
```

An important thing is I want to keep initial assets building logic form download logic, so `stock_df`  and `adj_factor` can't accept an input.
Since I don't see any example to chain the `download_stock` job and `stock_adj_df` asset .
Can you give me an example ?

**Comment by tacastillo** [comment]
Hi!

For your particular case, have you thought of turning that job into graph and a graph-backed asset?

[Graph-backed assets](https://docs.dagster.io/concepts/assets/graph-backed-assets) are a way to turn a series of ops that create an asset into an asset that Dagster can recognize.

Also, I notice that you're using the `create_shell_command_op` command. Have you thought about not using the `dagster-shell` library and running your shell scripts as subprocesses? Here's an example of how you can run shell scripts and produce an asset: https://github.com/dagster-io/dagster/discussions/15694

**Comment by eromoe** [comment]
A keypoint is I want to make 3 jobs seperated:
1. ops job : only download ( don't create asset) 
2. assets job: only transform data
3. chained 1&2 job : download and transform data


if  `turn a series of ops that create an asset`, every download  job create a asset .
And I want to keep code base as small as possible , don't tend to duplicate same logic in different place.
So the most idea way is create a new job, which connect job 1 and job 2 , to be job 3. 

I wonder if this is possible .

**Reply by sryza** [reply]
> ops job : only download ( don't create asset)

when you say "don't create asset", you're downloading it to somewhere, right? why not consider the location that you're downloading it to an asset?

**Reply by eromoe** [reply]
Because download/read&write/retry/partition logic is written in that lib.  Change the logic to dagster is not worth.
And I don't want every download create an asset, for now. 
I only use dagster to control download schedule, and almost manually trigger the downstream, it is a ml pipeline in develop.

**Reply by shenrun** [reply]
@eromoe 
I have a very similar use case as you and face the same problems. I am wondering whether you have figured out any solution or alternatives? I would really appreciate your suggestions.
